src/libsyntax/parse/parser.rs

```diff
@@ -3426,40 +3426,52 @@ impl<'a> Parser<'a> {
     }
 
     /// Parse the fields of a struct-like pattern
-    fn parse_pat_fields(&mut self) -> PResult<'a, (Vec<codemap::Spanned<ast::FieldPat>>, bool)> {
+    fn parse_pat_fields(&mut self)
+                        -> PartialPResult<'a, (Vec<codemap::Spanned<ast::FieldPat>>, bool)> {
         let mut fields = Vec::new();
         let mut etc = false;
         let mut first = true;
+
         while self.token != token::CloseDelim(token::Brace) {
             if first {
                 first = false;
             } else {
-                self.expect(&token::Comma)?;
+                self.expect(&token::Comma).map_err(|err| ((fields.clone(), etc), err))?;
                 // accept trailing commas
                 if self.check(&token::CloseDelim(token::Brace)) { break }
             }
 
-            let attrs = self.parse_outer_attributes()?;
+            let attrs = self.parse_outer_attributes()
+                .map_err(|err| ((fields.clone(), etc), err))?;
             let lo = self.span;
             let hi;
 
             if self.check(&token::DotDot) {
```

Comment:
The only thing this PR needs to do is to replace this `self.check(&token::DotDot)` with `self.check(&token::DotDot) || self.token == token::DotDotDot`, and report a non-fatal error (`self.span_err(...)`) if it's actually a `DotDotDot`.

Reply:
I see ...

(I might claim that the extra machinery in this PR is still potentially useful for taking successfully-parsed field patterns into account when there's some other kind of syntax error ... but maybe that can be a separate, future PR _if_ I find this `PartialPResult` device being useful elsewhere.)